### PR TITLE
Remove alternate file stream code from non-Windows

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
AlternateStreams support relies on pinvoke to win32 apis which don't work

on non-Windows and produces errors about not loading a dll in the case of copy-item

Fix https://github.com/PowerShell/PowerShell/issues/4542

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs]: `Unblock-File` is excluded from Unix at all so we don't need to exclude the code here.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Please add a check that `${testPath}:Stream` is a file name.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Extra line.

**Review by iSazonov — COMMENTED** [review]
Leave a comment

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Unnecessary change of line.

File: src/System.Management.Automation/utils/ClrFacade.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/utils/ClrFacade.cs]: Should this block be surrounded by #if !UNIX?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Should this be ```Test-Path``` instead of ```Test_path```

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs]: That file is still being compiled on Unix

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Doesn't `Get-Content` implicitly cover that?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: The change of grouping of the variables makes it a bit more readable to me

File: src/System.Management.Automation/utils/ClrFacade.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/utils/ClrFacade.cs]: Some of this code is unnecessary on UNIX so I'll wrap it

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Yes, odd that this test passed

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs]: I think we should exclude them in csproj.

Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs]: I'm fine with that but then we have effectively conditional compilation in both the source and the csproj files

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Will remove

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: I guess if the Content cmdlets were doing something funny it could mask an issue, I'll add a check

**Review by iSazonov — DISMISSED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (adityapatwardhan) [inline src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: Should we add ```#if !UNIX``` to the start of the file like: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/utils/tracing/EtwActivity.cs#L1

That seems more discoverable to me when doing code reviews or debugging.

**Review by adityapatwardhan — DISMISSED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: That's what I had originally.  Did a search and I don't see other cases of conditional compilation, so I think it's more consistent to use a whole file `#ifdef`.  @iSazonov ok?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs
Comment (dantraMSFT) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs]: Should we rename this to UnblockFile.Windows.cs? 

---
In reply to: [133130873](https://github.com/PowerShell/PowerShell/pull/4567#discussion_r133130873) [](ancestors = 133130873)

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: This #if/#endif block is cumbersome to read.  Suggest instead:
if (result != null)
{
#if UNIX
     WriteItemObject(...)
#else
    original code
 }

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: As above, the code is easier to read if you use a #if UNIX and contain the UNIX lines completely.  Its a simple if/else with a single statement for UNIX but  looks unwieldy with the current #if/#else placement.

**Comment by dantraMSFT** [comment]
>                     if (isAlternateDataStream)

Should this also be conditioned on !UNIX?

---
Refers to: src/System.Management.Automation/namespaces/FileSystemProvider.cs:4270 in 7209973. [](commit_id = 720997351e1de1cf26593cfe0e95699d8b3b9eb3, deletion_comment = False)

**Comment by dantraMSFT** [comment]
>             return Platform.NonWindowsIsSameFileSystemItem(pathOne, pathTwo);

Shouldn't there be a Platform.IsSameFileSystemItem?

---
Refers to: src/System.Management.Automation/namespaces/FileSystemProvider.cs:8274 in 7209973. [](commit_id = 720997351e1de1cf26593cfe0e95699d8b3b9eb3, deletion_comment = False)

File: src/System.Management.Automation/utils/ClrFacade.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/utils/ClrFacade.cs]: it appears you're still hitting alternate data streams on unix.
ClrFacade:  MapSecurityZoneWithoutUrlmon -> ReadFromZoneIdentifierDataStream

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: On UNIX, _stream will be null causing a ArgumentNullException from new StreamReader at line 874.  At best, it will be misleading. Consider throwing NotSupportException on UNIX in the associated constructors

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: I think it's technically better to exclude files in csproj files. But currently we don't use a lot of opportunities of MSBuild and also we are cleaning up FullCLR files and respectively remove "Compile Remove" from csproj files. So it is Ok to use  #ifdef .

/cc @daxian-dbw  Could you comment too?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: The `WriteItemObject` code applies to both UNIX and !UNIX.  If I move that bit of code up, I'd need more `#ifdefs` making it less readable

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I can move some code around so the UNIX code is in one place and the !UNIX code is also together.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: makes sense, will change

**Comment by SteveL-MSFT** [comment]
@dantraMSFT I addressed your feedback.  Regarding `Platform.IsSameFileSystem` we should probably have a single API that is platform agnostic, but that's out of scope of this PR.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Actually, if you don't mind duplicating the WriteItemObject line, it can be a pure #if #else block.  Your call.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: I think these exceptions need to be explicit in indicating alternate data streams are not supported. That's why I suggested an exception at ctor time; an exception that indicates streamName is not supported on UNIX.

File: src/System.Management.Automation/utils/ClrFacade.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/utils/ClrFacade.cs]: SecurityZone is a windows concept which I think shouldn't be applied when running on Unix plats. We shouldn't have `#if UNIX/endif` in this method or the methods exclusively used by this method, instead, we probably should analyze the call sites of this method to make sure this method doesn't run on unix plats, or better, enclose this method within `#if !UNIX/endif`.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I think the fix should be to make the dynamic parameters related to `Stream` operation unavailable in FileSystemProvider. Otherwise, it's broken user experience as I can use the `-Stream` parameter but it doesn't behave as expected.

**Review by daxian-dbw — DISMISSED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (JamesWTruher) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: yes, it would be better to not have the `-stream` parameter if it's not supported. It may also be cleaner to do it this way

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: I think it's better to make sure `CreateStreams` is never used for an alternate stream instead of inserting PlatformNotSupportedException.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Once we stop providing the `-Stream` dynamic parameter for FileSystemProvider on Unix, we don't need to add so many `#if UNIX/endif` blocks in this file because those codes will not be hit. I think that will make the code more readable and easy to maintain.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: if you approach this by removing the dynamic parameter on ! windows, we should probably have an explicit test for the parameter - windows => parameter present ; ! windows => parameter absent

File: src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj]: I don't really have a preference although `csproj` approach seems more discoverable. However in this case, I don't think `Unblockfile.cs` should be completely excluded from Unix build as it will increase the number of `#if unix/endif` blocks in places where the code from this file is used, which will impact readability.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (adityapatwardhan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Closed.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: The PSCopyToSessionHelper script and PSCopyFromRemoteSession script for remote file copy contains support for alternate streams.  Ideally we would have versions of this script for Linux platforms that remove that support.  But I am not sure it should be part of this change.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: This method can be made private access since it is only used here.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (PaulHigin) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Likewise, should be private access.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I think the other case I was fine with a bit of duplication, but prefer to leave this one as-is.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: Good idea, will change.

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: ditto

File: src/System.Management.Automation/namespaces/FileSystemContentStream.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemContentStream.cs]: see @daxian-dbw comment below as better way to address this

File: src/System.Management.Automation/utils/ClrFacade.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/utils/ClrFacade.cs]: There are some other checks this code does, however, upon closer inspection, I think they also don't apply to Unix so I'll `#ifdef` this code out on UNIX

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: So we still want the code compiled on UNIX?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: will add

File: src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs]: I believe CDXML cmdlets can work on Linux via OMI

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I think that code cleanup should be part of a different PR.  Created https://github.com/PowerShell/PowerShell/issues/4601

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: will do

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: will do

File: test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1]: Need to check if there are already tests using -stream otherwise I'll add

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: If that makes the code cleaner overall, then I'm fine to have the code compiled on Unix.

Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Looking at the existing code, the FileSystemProvider has a ton of `#if UNIX` and `#if !UNIX` already because the filesystems are inherently different (beyond just NTFS streams).  I think refactoring this code is an exercise best left for the future.

**Comment by SteveL-MSFT** [comment]
I believe all feedback has been addressed

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (dantraMSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Shouldn't this be in the #if !UNIX block?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: The alternate streams can be specified using `-Stream` or an inline stream syntax `"file::stream"`, the check of the inline stream syntax is done in 7 places in `FileSystemProvider.cs` but some of them are not within `#if !UNIX` block.

Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: And when running `Get-Content -Path file::stream` on Linux, it probably should throw an error message about "alternate stream is not supported" instead of working unexpectedly or fail with other errors.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (daxian-dbw) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: I thought about it more and now have this question: which the following experience is better -- copy `Get-Content -Path <file> -Stream <stream>` to Linux powershell and see the error `"A parameter cannot be found"`, or `"Alternate stream is not supported on this platform"`? I think the latter message is better. To achieve this, maybe we can make the setter of the `Stream` property throw `PlatformNotSupportedException` on non-windows.

As for `#if UNIX/endif` blocks, IMHO, we don't necessarily need to have all codes that only apply to a specific platform guarded by the compile definition. I think it's OK to keep such codes around even for other platforms, and the guideline should be readability and maintainability. I still feel it's better to avoid most of the `#if UNIX` blocks by controlling the entry points:
1. Make sure `-Stream` not available on Unix platforms
2. Throw error when the inline stream syntax is used for `-Path` on Unix platforms
3. Skip the alternate stream discovery codes during remote copy (both `PerformCopyFileFromRemoteSession` and `PerformCopyFileToRemoteSession`) on Unix platforms.

After controlling those entry points (and put some comments for sure), it's safe to know that the alternate stream related code paths will not be hit on Unix platforms.

File: assets/Product.wxs
Comment (TravisEz13) [inline assets/Product.wxs]: something isn't correct in your PR.  Already committed changes are in the diff.

**Review by TravisEz13 — DISMISSED** [review]
Please work on correctly resolving merge/rebase issue.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: With the current changes, you don't get any error using `copy-item` going from Linux->Windows nor Windows->Linux

I created a new issue https://github.com/PowerShell/PowerShell/issues/4633 to clean up the code later since filesystemprovider.cs already has 21 `#if` statements before I made any changes

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: yes, will fix

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: On Unix, a colon is completely allowed in a filename, so we want that to continue working as just a filename.  Tests were added to make sure this works.

I'll fix other cases of using the `file:stream` syntax.

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: will fix

Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: will fix

**Comment by TravisEz13** [comment]
It looks like the email on your latest commit is not associated with your github account.

**Comment by SteveL-MSFT** [comment]
@TravisEz13 fixed

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs
Comment (lzybkr) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: Which is more likely - somebody copying a script from Windows to Linux that uses `-Stream`, or somebody on Linux using parameter completion, sees `-Stream` as a possible option, and wastes time trying to learn what it does?

I'm thinking it's better overall to not suggest parameters on platform if they aren't actually supported.

Comment (SteveL-MSFT) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs]: `Copy-Item` doesn't expose a `-Stream` param at all.  In the case of copying a NTFS file that contains alternate streams, those streams don't exist on the Linux destination.

**Comment by SteveL-MSFT** [comment]
@TravisEz13 @daxian-dbw feedback addressed

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**
